### PR TITLE
Polishing StreamDefinition service

### PR DESCRIPTION
**Pull request by ilayaperumalg**
Add app metadata configuration for inbound/outbound port mapping

StreamDefinition controller changes to return application definitions on a specific REST endpoint

Add Redact methods for stream definition sanitization

Update tests

### CODE REVIEW

File: spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/DefaultStreamDefinitionService.java [hunk 1]
```diff
@@ -53,4 +58,71 @@ public StreamNode parse(StreamDefinition streamDefinition) {
...
+
+		return dsl;
+	}
+
+	@Override
+	public String redactDsl(StreamDefinition streamDefinition) {
+		return this.constructDsl(streamDefinition.getDslText(), StreamDefinitionServiceUtils.sanitizeStreamAppDefinitions(this.getAppDefinitions(streamDefinition)));
+	}
+
+	public static String unescape(String text) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/DefaultStreamDefinitionService.java:124]: public static method is probably not a good idea.

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java [hunk 2]
```diff
@@ -234,6 +282,34 @@ else if ("_ROOT_GROUP_".equals(group.getId())) {
 		}
 	}
 
+	public Map<String, Set<String>> listPortNames(Archive archive) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java:285]: Maybe private or protected as `resolvePortNamesFromContainerImage` is private

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java [hunk 3]
```diff
@@ -234,6 +282,34 @@ else if ("_ROOT_GROUP_".equals(group.getId())) {
...
+						.delimitedListToStringArray(properties.getProperty(CONFIGURATION_PROPERTIES_INBOUND_PORTS), ",", " ")));
+				portsMap.put("inbound", inboundPorts);
+				outboundPorts.addAll(Arrays.asList(StringUtils
+						.delimitedListToStringArray(properties.getProperty(CONFIGURATION_PROPERTIES_OUTBOUND_PORTS), ",", " ")));
+				portsMap.put("outbound", outboundPorts);
+			}
+			return portsMap;
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Exception trying to list configuration properties for application " + archive,
```
Comment (jvalkeal) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java:308]: Shouldn't this be `AppMetadataResolutionException`

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java [hunk 4]
```diff
@@ -147,6 +152,27 @@ public BootApplicationConfigurationMetadataResolver(ClassLoader parent,
 		return Collections.emptyList();
 	}
 
+	@Override
+	public Map<String, Set<String>> listPortNames(Resource app) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java:156]: nitpick: Easy to tidy up whole function.

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java [hunk 5]
```diff
@@ -180,12 +206,34 @@ private boolean isDockerSchema(URI uri) {
...
+			return Collections.emptyMap();
+		}
+
+		String encodedMetadata = labels.get(CONTAINER_IMAGE_CONFIGURATION_METADATA_LABEL_NAME);
+		if (!StringUtils.hasText(encodedMetadata)) {
+			return Collections.emptyMap();
+		}
+
+		try {
+			//todo: add port mappings from container image config
```
Comment (jvalkeal) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java:223]: Do we leave this unimplemented?

Comment (ilayaperumalg) [reply spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/BootApplicationConfigurationMetadataResolver.java:223]: No, I'll have to work on this post M2.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/AppRegistryController.java [hunk 6]
```diff
@@ -189,6 +190,21 @@ private DetailedAppRegistrationResource getInfo(ApplicationType type,
 		for (ConfigurationMetadataProperty property : properties) {
 			result.addOption(property);
 		}
+		Map<String, Set<String>> portsMap = this.metadataResolver.listPortNames(appRegistryService.getAppMetadataResource(registration));
+		if (portsMap != null && !portsMap.isEmpty()) {
+		for (Map.Entry<String, Set<String>> entry: portsMap.entrySet()) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/AppRegistryController.java:195]: wrong indentation, weird that checkstyle didn't fail.

**Comment by jvalkeal** [comment]
Merged per 582fb52deb0031e520e368f5df7520843857e021

**Pull request closed without merging**